### PR TITLE
Support for Flutter Web

### PR DESCRIPTION
**Pull request by rodydavis**
- Uses Cookies to store Json
- Dynamic Imports for Path Provider and only used on mobile and desktop

### CODE REVIEW

**Comment by lesnitsky** [comment]
@AppleEducate I was thinking of using actual browser localstorage, so this package is just a "proxy" in the web. Using cookies works, but localstorage has its own advantages (like ability to listen for changes in different tab)

Also, path provider related stuff should be refactored in any case, since `path_provider` implementation is available in flutter-desktop-embedding repo

**Comment by rodydavis** [comment]
Yeah makes sense. Good point. Ill see what i can do

**Comment by lesnitsky** [comment]
Maybe let me do the proper refactor first for existing platforms... so that we won't have too much of a conflicts
Also, do you know if a proper plugin system for web exists? `flutter create` doesn't seem to create a `web` folder

**Comment by rodydavis** [comment]
Ok Refactored to do proper web local storage

- also added stream builder
- abstracted directory path

If you are on the master channel flutter create will create the web folder. And web plugins do exist already.

**Comment by rodydavis** [comment]
This will now work across desktop, mobile and web! Works pretty great too! Tested in chrome and survives after a restart

**Comment by rodydavis** [comment]
The stream is nice so that you can use it in a stream builder now and stateless widgets

File: .vscode/launch.json [hunk 1]
```diff
@@ -0,0 +1,14 @@
+{
```
Comment (Cretezy) [inline .vscode/launch.json:1]: You can ignore .vscode and .idea in your global gitignore

https://gist.github.com/subfuzion/db7f57fff2fb6998a16c

File: lib/localstorage.dart [hunk 2]
```diff
@@ -1,17 +1,18 @@
-import 'dart:io';
 import 'dart:async';
 import 'dart:convert';
+
 import 'package:flutter/foundation.dart' show ValueNotifier;
-import 'package:path_provider/path_provider.dart';
+
+import 'src/directory/directory.dart';
```
Comment (Cretezy) [inline lib/localstorage.dart:6]: Use `package:` here

File: lib/src/directory/io.dart [hunk 3]
```diff
@@ -0,0 +1,117 @@
+import 'dart:async';
+import 'dart:convert';
+import 'dart:io';
+
+import 'package:path_provider/path_provider.dart';
+
+import '../impl.dart';
```
Comment (Cretezy) [inline lib/src/directory/io.dart:7]: `package:` here too

File: lib/src/directory/unsupported.dart [hunk 4]
```diff
@@ -0,0 +1,55 @@
+import 'dart:async';
+
+import '../impl.dart';
```
Comment (Cretezy) [inline lib/src/directory/unsupported.dart:3]: And here

File: lib/src/directory/web.dart [hunk 5]
```diff
@@ -0,0 +1,92 @@
+import 'dart:async';
+import 'dart:convert';
+import 'dart:html' as html;
+
+import '../impl.dart';
```
Comment (Cretezy) [inline lib/src/directory/web.dart:5]: And here

File: lib/src/impl.dart [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+abstract class LocalStorageImpl {
+  LocalStorageImpl(this.fileName, [this.path]);
+
+  final String path, fileName;
+
+  Stream<Map<String, dynamic>> get stream;
+
+  void dispose();
+
+  Future init([Map<String, dynamic> initialData]);
```
Comment (Cretezy) [inline lib/src/impl.dart:12]: All these `Future` should have a return type. By default it is `dynamic`, you probably want `void`

File: lib/localstorage.dart (hunk 2 above)
Comment (rodydavis) [reply lib/localstorage.dart:6]: The reason why you have to use this is so it works on the web. You need conditional imports so it pulls in the right code at compile time

File: lib/src/directory/io.dart (hunk 3 above)
Comment (rodydavis) [reply lib/src/directory/io.dart:7]: It’s needed for web support

Comment (Cretezy) [reply lib/src/directory/io.dart:7]: I'm saying the import should be

`import 'package:flutter_localstorage/src/impl.dart`

As per the Dart recommendation

File: lib/localstorage.dart (hunk 2 above)
Comment (Cretezy) [reply lib/localstorage.dart:6]: See other comment. You should use the package-style import

**Comment by lesnitsky** [comment]
@AppleEducate I'm going to do some polishing before release

**Pull request merged**